CA SAbowitz

lepidopterane-atsmith opened the pull request:
Here it is, the first finished version of the comprehensive JUnit
Tester for CA. Please let me know if you have any
questions/concerns/whatever.

lepidopterane-atsmith committed:
CA Tester V1.0.0

This is incomplete but here’s a look at what I’m working on. It’s based
on the BasicControllerTest. Please read my note before testing (though
this is not ready to run yet).

lepidopterane-atsmith committed:
CA Tester V1.0.1

Still incomplete, but this commit reflects that I’m 70% there.

lepidopterane-atsmith committed:
CA Tester V1.0.2

Here it is, the first finished version of the comprehensive JUnit
Tester for CA. Please let me know if you have any
questions/concerns/whatever.

lepidopterane-atsmith committed:
CA Tester V1.0.3

Whoops, forgot to actually plug this into TestRunner!

lepidopterane-atsmith committed:
CA Test Suite V1.0.4

Forgot some things in ConversationAccessServerTest (CAST). Also
demonstrates what I’ve been working on lately.

lepidopterane-atsmith committed:
CA Test Suite V1.0.5

Trying to ungank this project so it doesn’t confuse Eclipse.

lepidopterane-atsmith committed:
CA Test Suite V1.0.6

Still polishing some odds and ends…

lepidopterane-atsmith committed:
CA Test Suite V1.0.7

There, that’s better.

lepidopterane-atsmith committed:
CA Test Suite 1.0.8

Further cleaning.

lepidopterane-atsmith committed:
CA Test Suite 1.0.9

Now tests the different nonpersistent storage mechanism I’ve made for
this feature.

lepidopterane-atsmith commented:
Note: To see my test of the storage system I implemented in action, run TestRunner.java .

aslo commented on .classpath:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
```
Since this is an eclipse config file and is particular to your development environment, you should not commit it. It's probably best to add this filename to the `.gitignore` file.

lepidopterane-atsmith replied:
how

aslo replied:
Sorry I didn't see this comment until now. You already added it to the .gitignore file, so now you just need to run `git rm --cached .classpath` to tell git to "forget" about the file. 

https://stackoverflow.com/questions/1274057/how-to-make-git-forget-about-a-file-that-was-tracked-but-is-now-in-gitignore

aslo commented on .project:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
```
This is also an eclipse config file - you should add it to the .gitignore file.

aslo replied:
You should `git rm --cached .project` for this file as well.

aslo commented on src/codeu/chat/server/Model.java:
```diff
@@ -140,10 +146,16 @@ public InterestSet getInterestSet(Uuid id){
   }
+  
+  public HashMap<Uuid, Byte> getPermissionMap(ConversationHeader c){
+	  return (HashMap<Uuid, Byte>) authority.get(c.title);
```
In general, you should avoid casting where possible, because it introduces the possibility of `ClassCastExceptions` into your program, which can be hard to avoid as your code evolves over time.

You actually don't need to do this here - you can just change the the `authority` map to have type `HashMap<String, HashMap<Uuid, Byte>>`. Then `authority.get(...)` will return an object with type `HashMap<Uuid, Byte>`, no casting necessary!

aslo commented on test/codeu/chat/server/PermissionMapTest.java:
```diff
@@ -0,0 +1,51 @@
+	  
+	  HashMap<Uuid, Byte> accessMap = model.getPermissionMap(chatA);
+
+	  assertFalse(
```
Rather than the double negative, why not `assertTrue(... , accessMap.containsKey(p1.id))`?

aslo commented on test/codeu/chat/server/PermissionMapTest.java:
```diff
@@ -0,0 +1,51 @@
+	  HashMap<Uuid, Byte> accessMap = model.getPermissionMap(chatA);
+
+	  assertFalse(
+		  "Check that the correct map was accessed",
```
FYI, assertion messages like this are nice to include, but should be thought of as purely optional. Be mindful of the amount of time/brainpower that you're spending writing/updating these messages over time!

aslo commented on test/codeu/chat/server/PermissionMapTest.java:
```diff
@@ -0,0 +1,51 @@
+
+	  assertFalse(
+		  "Check that the correct map was accessed",
+		  !accessMap.containsKey(p1.id));
```
It would also be good to make an assertion about the value in the map. That is, after all, the most important part!

aslo commented on test/codeu/chat/server/ConversationAccessServerTest.java:
```diff
@@ -0,0 +1,465 @@
+
+    final User user = controller.newUser("user");
+
+    assertFalse(
```
Consider using `assertNotNull` instead. It makes it a bit easier as a programmer to read and understand why tests have failed.

aslo replied:
Beyond asserting that the User is not null, it would also be good to make some assertions about the user's properties as well - i.e. is the user's name set as you expect? Does it have a valid UUid? These types of assertions are good because they test even more of the controller's logic.

aslo commented on test/codeu/chat/server/ConversationAccessServerTest.java:
```diff
@@ -0,0 +1,465 @@
+    accessMap.put(user.id, memberByte);
+    controller.updatePermissionMap(conversation, accessMap);
+    
+    assertFalse(
```
I'd opt for `assertTrue` instead - it's easier to reason about than a double negative.

aslo commented on test/codeu/chat/server/ConversationAccessServerTest.java:
```diff
@@ -0,0 +1,465 @@
+  }
+  
+  @Test
+  public void testAccessSecurity(){
```
This method is very long - about 200 lines. Whenever you find yourself writing a method of this length, consider ways that you might refactor your code to make it smaller so that it's easier to understand. 

When it comes to unit tests specifically, it's considered good practice to write many small, concise test methods, each of which should ideally only test one thing. (To give you an idea, some people feel that one should only have a single assertion per test method.  I'm not that dogmatic about it, but it's a nice goal to aim for when writing tests).

aslo commented on test/codeu/chat/server/ConversationAccessServerTest.java:
```diff
@@ -0,0 +1,465 @@
+  // The following test simulates a more realistic situation where different users have
+  // different roles in the conversations they contribute to.
+  @Test
+  public void multiConversationTest(){
```
Same here - you should think about ways to break this method into many smaller tests. I'm happy to suggest some ideas as well.

aslo reviewed (changes requested):
This is looking great! Major points for taking on testing, which is a crucial part of software development and is new to you!

This PR is a little hard to review since the methods you're testing are not yet implemented. In this sort of situation, where you're testing code that hasn't yet been written, it's usually best to just implement placeholders for the methods you're testing. (Just empty methods that either return null, or throw a RuntimeException, etc. Usually it's also good to add a comment saying that these methods are just placeholders). That way, you can at least run the tests, even if they fail - which is better than test code that doesn't compile.

lepidopterane-atsmith commented:
@aslo - Thanks so much for the tips on test development! I also pushed a storage system that's tested using the current version of TestRunner in this pull and it's based on this series of thoughts in the project group chat vv
![screen shot 2017-07-17 at 4 27 01 pm](https://user-images.githubusercontent.com/23276976/28288419-1ea46756-6b0d-11e7-8c28-15142305b506.png)

What do you think of it as a different way to store access maps by conversation?

aslo commented:
@lepidopterane-atsmith No problem! That depends. What exactly are you trying to accomplish with that change?

lepidopterane-atsmith commented:
I wanted to establish that this way of storage was an efficient way of storing access maps for the feature, kind of like Interest System's storage in Model. I was hoping to establish something that could be securely accessed/updated speedily in a format like HashMap<Conversation, HashMap<Uuid, byte>>.  In the design doc, I argue that storing authority this way would mean "less traipsing around for m-ban and other m-commands like that because they all take place in the current conversation." Hope this helps!

lepidopterane-atsmith commented:
Actually, nvm, I thought you didn't address something but you did, sorry

lepidopterane-atsmith committed:
CA Test Suite V1.1.0

Polished some things. Refactored testAccessSecurity() into 16 separate
tests.

lepidopterane-atsmith committed:
CA Test Suite V1.1.0.5

Further polishing...

lepidopterane-atsmith committed:
CA Test Suite V1.1.1

No more tests containing 200 lines and 20 asserts! These are refactored
now.

aslo commented on test/codeu/chat/server/ConversationAccessServerTest.java:
```diff
@@ -0,0 +1,535 @@
+        "conversation",
+        user.id);
+
+    assertFalse(
```
It's cleaner to just use `assertNotNull` here.

aslo commented on test/codeu/chat/server/ConversationAccessServerTest.java:
```diff
@@ -0,0 +1,535 @@
+        "Check that conversation has a valid reference",
+        conversation == null);    
+        
+    assertNotNull(
```
You should go a step further here and assert that the contents of the permission map that was returned by `controller.getPermissionMap(conversation)` are correct.

aslo commented on test/codeu/chat/server/ConversationAccessServerTest.java:
```diff
@@ -0,0 +1,535 @@
+    
+    assertTrue(
+        "Check that the allowances can be updated",
+        controller.getPermissionMap(conversation).size()== 1);
```
Again, you should go a step further, and examine the contents of the permission map. Here, you're testing that the permission map has an additional entry, but you should also make sure that the new entry is correct.

aslo commented on test/codeu/chat/server/ConversationAccessServerTest.java:
```diff
@@ -0,0 +1,535 @@
+	ConversationHeader conversation = spawnTestConversation();
+    HashMap<Uuid, Byte> accessMap = controller.getPermissionMap(conversation);
+  
+    assertTrue(
```
When you're comparing two items for equality in your test, I'd recommend using the `assertEquals(a,b)` method, instead of `assertTrue(a == b)` or `assertTrue(a.equals(b))`. It's a bit cleaner, but more importantly, if the test fails, the error message will be much more readable. I recommend making this change for all instances of this in this test.

aslo commented on test/codeu/chat/server/ConversationAccessServerTest.java:
```diff
@@ -0,0 +1,535 @@
+		"Check automatic membership for newcomers to chat",
+		controller.checkMembership(member.id, conversation));
+		    	
+	assertTrue(
```
This is identical to the above assertion - you should delete one.

aslo commented on test/codeu/chat/server/ConversationAccessServerTest.java:
```diff
@@ -0,0 +1,535 @@
+	  
+	assertTrue(
+		"Check that illegal cases of addOwner do nothing",
+		accessMap.equals(controller.getPermissionMap(conversation)));  
```
This `accessMap` variable is never initialized, so this won't compile.

aslo commented on test/codeu/chat/server/ConversationAccessServerTest.java:
```diff
@@ -0,0 +1,535 @@
+	  
+	assertTrue(
+		"Check that illegal cases of demoteOwner do nothing",
+		accessMap.equals(controller.getPermissionMap(conversation)));
```
This variable is initialized below this line - please fix, otherwise this won't compile.

aslo commented on test/codeu/chat/server/PermissionMapTest.java:
```diff
@@ -0,0 +1,60 @@
+		  "Check that the correct value was stored in the map",
+		  accessMap.get(p1.id).byteValue()==creatorByte);
+	  
+	  print();
```
What is this method?

lepidopterane-atsmith replied:
That's a scrap, sorry about that!

aslo commented on test/codeu/chat/server/PermissionMapTest.java:
```diff
@@ -0,0 +1,60 @@
+import codeu.chat.common.User;
+import codeu.chat.util.Uuid;
+
+public class PermissionMapTest {
```
Why use a separate class for these tests? Do you think you could just merge this with `ConversationServerAccessTest`? Both classes seem to be testing the same things.

aslo reviewed (changes requested):
Making great progress! I left a few more comments to address. Also, don't forget to address the remaining comments from my last review. Thanks!

lepidopterane-atsmith committed:
CA Test Suite V1.1.2

Addressed Alex’s most recent comments and also incorporated more of the
actual commands into testing.

lepidopterane-atsmith committed:
Model now implements AuthorityModel!

Glad that’s done!

lepidopterane-atsmith committed:
CA Test Suite now operational

Successfully grafted all of Yuhang’s code into my environment. Now we
can put this tester to good use!

lepidopterane-atsmith committed:
CA TestSuite V1.1.3

Fixed some errors under my responsibility.

lepidopterane-atsmith committed:
CA V1.0.0 I1

first intermediary version of the final product while I try to figure
out how to get Yuhang’s branch

lepidopterane-atsmith committed:
CA V1.0.1

I think everything works now?

aslo commented on src/codeu/chat/common/BasicController.java:
```diff
@@ -58,6 +59,9 @@
+  // Shuttles in an Interest Set updated from the client-end.
   void updateInterests(Uuid id, InterestSet intSet);
+  
+  // Yuhang, I'll let you describe this one.
```
Make sure you either delete this comment or fill it in.

aslo commented on src/codeu/chat/server/AuthorityModel.java:
```diff
@@ -0,0 +1,23 @@
+/* Original interface by Yuhang Liao, 
+ * annotation and imports by Sarah Abowitz.
+ * 
+ * Hey idk if the new v of this is worth it
```
You should delete comments like these - keep them to only what's informative to readers of your code.

aslo commented on src/codeu/chat/server/Model.java:
```diff
@@ -135,15 +146,133 @@ public void add(Message message) {
+  public boolean isMember(ConversationHeader conversation,Uuid targetUser){
+    HashMap<Uuid, Byte> accessMap = this.getPermissionMap(conversation);
+    if(!accessMap.containsKey(targetUser)){
+  	  accessMap.put(targetUser, USER_TYPE_MEMBER);
```
This looks like a bug to me. If there's a new user who isn't a member of a conversation, checking their membership status should not make them a member.

aslo commented on src/codeu/chat/server/Model.java:
```diff
@@ -135,15 +146,133 @@ public void add(Message message) {
+	ConversationHeader chat = convos.first(conversation.root());
+	HashMap<Uuid, Byte> accessMap = this.getPermissionMap(chat);
+	if(!accessMap.containsKey(targetUser)){
+	  accessMap.put(targetUser, USER_TYPE_MEMBER);
```
Same issue here. I don't think you should be adding membership status when checking if a user is a member.

aslo commented on src/codeu/chat/server/Model.java:
```diff
@@ -135,15 +146,133 @@ public void add(Message message) {
+	ConversationHeader chat = convos.first(conversation.root());
+	HashMap<Uuid, Byte> accessMap = this.getPermissionMap(chat);
+	if (!accessMap.containsKey(targetUser)){
+	  this.isMember(conversation, targetUser);
```
What's the purpose of calling this method here?

aslo commented on src/codeu/chat/server/Model.java:
```diff
@@ -135,15 +146,133 @@ public void add(Message message) {
+  public boolean isOwner(ConversationHeader conversation,Uuid targetUser){
+	HashMap<Uuid, Byte> accessMap = this.getPermissionMap(conversation);
+	if (!accessMap.containsKey(targetUser)){
+	  this.isMember(conversation, targetUser);
```
What's the purpose of calling this method here?

lepidopterane-atsmith replied:
Every similar method call usage like this was part of the idea that anyone who wasn't on the roster with a status already in a chat would be assigned to member status as a default. Seeing that this phase of thinking wasn't the best idea, I'll be updating these methods appropriately.

aslo commented on src/codeu/chat/server/Model.java:
```diff
@@ -135,15 +146,133 @@ public void add(Message message) {
+  public boolean isCreator(ConversationHeader conversation,Uuid targetUser){
+	HashMap<Uuid, Byte> accessMap = this.getPermissionMap(conversation);
+	if (!accessMap.containsKey(targetUser)){
+		  this.isMember(conversation, targetUser);
```
What's the purpose of calling this method here?

aslo commented on src/codeu/chat/server/Model.java:
```diff
@@ -135,15 +146,133 @@ public void add(Message message) {
+	ConversationHeader chat = convos.first(conversation.root());
+	HashMap<Uuid, Byte> accessMap = this.getPermissionMap(chat);
+	if (!accessMap.containsKey(targetUser)){
+		  this.isMember(conversation, targetUser);
```
What's the purpose of calling this method here?

aslo reviewed (changes requested):
This is very close! Just a few comments for you to address.

lepidopterane-atsmith committed:
Replaced placeholder with important message

lepidopterane-atsmith committed:
Deleted useless comment in AuthorityModel

lepidopterane-atsmith committed:
Aborted member-default for non-rostered individuals

lepidopterane-atsmith committed:
CA V1.0.2

lepidopterane-atsmith committed:
CA V1.0.3

One of the tests was inaccurate bc we decided to nix
newcomer/membership defaults. That’s fixed now.
